src/main/java/seedu/programmer/ui/CommandBox.java

```diff
@@ -67,16 +71,25 @@ private void handleKeyPressed(KeyEvent event) {
         boolean downPressed = event.getCode() == KeyCode.DOWN;
         boolean isNotUpOrDown = !upPressed && !downPressed;
         boolean noCommandHistoryPresent = commandHistory.isEmpty();
-        boolean downAndAtLastCommand = downPressed && commandHistory.isCounterAtLast();
-        boolean upAndAtFirstCommand = upPressed && commandHistory.isCounterAtFirst();
+        boolean downAndAtLastCommand = downPressed && commandHistory.isAtLastIndex();
+        boolean upAndAtFirstCommand = upPressed && commandHistory.isAtFirstIndex();
+        boolean upAndAtLastCommand = upPressed && commandHistory.isAtLastIndex();
 
         // Do nothing if neither up nor down pressed, or no command history
         if (isNotUpOrDown || noCommandHistoryPresent) {
             return;
         }
 
-        if (downAndAtLastCommand || upAndAtFirstCommand) {
-            logger.info("We are already at the oldest or newest command -> show current command");
+        if (upAndAtLastCommand && lastCommandIsValid) {
+            // Remember the text field resets to "" after each valid command
+            commandTextField.setText(commandHistory.getCurrentCommand());
+            lastCommandIsValid = false;
+        } else if (downAndAtLastCommand && lastCommandIsValid) {
+            logger.info("We are already at the newest command -> show current command");
+            commandTextField.setText("");
```

Comment:
^there is still a bug now, to replicate error:
1. `show 1`
2. `show 2`
3. `show 3`
4. press down once --> nothing happens
5. press down again --> current command appears
6. then works as per normal

Suggestion:
1. Change line 89 to this line `commandTextField.setText(commandHistory.getCurrentCommand());`

Reply:
Good catch! How about removing line 90? Seems to work for me